### PR TITLE
Adjusts in SKIP and CANCEL [V4]

### PR DESCRIPTION
**Pull request by apahim**
**Goals**:

SKIP

- Will never run `setUp()`/`tearDown()`.
- Can be triggered using skip decorators.
- Can be triggered using `self.skip()`.
- `self.skip()` can still be used only in `setUp()`
- Start deprecating `self.skip()`

CANCEL:

- Will run `setUp()` and `tearDown()`.
- Can be triggered using `self.cancel()`.
- `self.cancel()` can be called anywhere, including in `setUp()` and `tearDown()`.

DOCS:
- Update documentation to consider the new CANCEL status.

---
v4:
- Removed the code that would not accept skip decorators to be used in `setUp()`. This topic will be internally discussed before come to github.

v3: #1878 
- Reordered commits so the deprecation `self.skip()` deprecation message reflects a working behavior.
- Make `self.cancel()` called in `setUp()` to run `tearDown()`.
- Marge the two commits handling `SKIP` in one commit only.
- Renamed `function.__skip__` to `function.__skip_test_decorator__`.
- Be explicit in the `True`/`False` test of `function.__skip_test_decorator__`.
- Selftests improvements to cover the changes in this PR.
- Change example test `skiponsetup.py` to reflect the deprecation process.

v2: #1874 
- Improve deprecation message to self.skip()
- Don't accept skip decorators in setUp() and tearDown()
- Accept self.cancel() in setUp() and tearDown()
- Rebase commits, putting the corresponding docs along with the code.

v1: #1872 
- Start the deprecation process for `self.skip()`.
- Don't execute `setUp()` and `tearDown()` when using skip decorators.
- Docs adjustments.

### CODE REVIEW

**Comment by apahim** [comment]
[Add support to "CANCEL" test execution](https://trello.com/c/viBJIEwI/941-add-support-to-cancel-test-execution)

File: avocado/core/test.py [hunk 1]
```diff
@@ -572,17 +573,18 @@ def _run_avocado(self):
...
-            raise exceptions.TestError(skip_illegal_msg)
+            cancel_test = details
         except:  # Old-style exceptions are not inherited from Exception()
             stacktrace.log_exc_info(sys.exc_info(), logger='avocado.test')
             details = sys.exc_info()[1]
             raise exceptions.TestSetupFail(details)
         try:
+            #TODO: This should be removed when this try/testMethod()
+            #      becomes an else/testMethod(). Then the TestCance
+            #      exception has to be raised in the setUp() `except`.
```
Comment (ldoktor) [inline avocado/core/test.py]: Well this is not yet decided so we don't actually know what happens... I'd just leave it on the one who takes this task to decide how is he going to implement it...

Comment (ldoktor) [inline avocado/core/test.py]: I mean there is a big room for optimization including the `__skip_test_decorator__` which can be tackled at the beginning, not polluting the main execution...

Comment (ldoktor) [inline avocado/core/test.py]: ... which means no actions are required, just leaving a message that the comment was probably not really essential...

File: docs/source/WritingTests.rst [hunk 2]
```diff
@@ -960,6 +960,10 @@ Avocado offers some options for the test writers to skip a test:
 Test ``skip()`` Method
 ----------------------
 
+.. warning:: `self.skip()` will be deprecated by the end of 2017.
```
Comment (ldoktor) [inline docs/source/WritingTests.rst]: Nitpick: on English lessons we learned that `by ...` means sometimes before but close to the $DATE and `on ...` means exactly on the $DATE. Shouldn't this be `at the end of 2017`? (not that it'd matters, I'm just wondering whether my English is up to date and how it works in the real life...

File: examples/tests/cancelonsetup.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env python
```
Comment (ldoktor) [inline examples/tests/cancelonsetup.py]: Well moving the `skip` to `cancel` would make it clearer, but this is also understandable...

Comment (ldoktor) [inline examples/tests/cancelonsetup.py]: Also you could probably remove the `skip_outside_setup` as it is a negative test which is not used anywhere...

File: docs/source/WritingTests.rst (hunk 2 above)
Comment (apahim) [inline docs/source/WritingTests.rst]: I meant close to the end of the year. Like a Christmas gift :)
But I can change it. What would you prefer?

File: avocado/core/test.py (hunk 1 above)
Comment (apahim) [inline avocado/core/test.py]: Sure, I can remove it.

File: docs/source/WritingTests.rst (hunk 2 above)
Comment (ldoktor) [inline docs/source/WritingTests.rst]: I don't mind the current, I'm curious whether I remember the rules correctly and what the native say about it as international English is sometimes different from the Oxford one...

**Comment by apahim** [comment]
@ldoktor, sure, tranks. V5 on the way.

**Pull request closed without merging**